src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -238,33 +252,36 @@ private void AddCompletionItems(List<MatchResult> list, CancellationToken cancel
 
-                using var _1 = PooledHashSet<string>.GetInstance(out var includedPreferredItems);
-                using var _2 = ArrayBuilder<MatchResult>.GetInstance(out var includedDefaults);
-                var unmatchedDefaults = _snapshotData.Defaults;
+                var includedPreferredItems = new ConcurrentSet<string>();
+                var includedDefaults = new ConcurrentDictionary<string, MatchResult>();
+
+                Enumerable.Range(0, _snapshotData.InitialSortedItemList.Count)
+                    .AsParallel()
+                    .WithCancellation(cancellationToken)
+                    .ForAll(CreateMatchResultAndProcessMatchingDefaults);
 
-                // Filter items based on the selected filters and matching.
-                var totalCount = _snapshotData.InitialSortedItemList.Count;
-                for (var currentIndex = 0; currentIndex < totalCount; currentIndex++)
+                PromoteDefaultItemsToPreferredState();
+
+                void CreateMatchResultAndProcessMatchingDefaults(int index)
                 {
-                    cancellationToken.ThrowIfCancellationRequested();
-                    var item = _snapshotData.InitialSortedItemList[currentIndex];
+                    var item = _snapshotData.InitialSortedItemList[index];
 
                     // All items passed in should contain a CompletionItemData object in the property bag,
                     // which is guaranteed in `ItemManager.SortCompletionListAsync`.
                     if (!CompletionItemData.TryGetData(item, out var itemData))
                         throw ExceptionUtilities.Unreachable();
 
                     if (filterHelper.ShouldBeFilteredOut(item))
-                        continue;
+                        return;
 
                     // currentIndex is used to track the index of the VS CompletionItem in the initial sorted list to maintain a map from Roslyn item to VS item.
                     // It's also used to sort the items by pattern matching results while preserving the original alphabetical order for items with
                     // same pattern match score since `List<T>.Sort` isn't stable.
-                    if (CompletionHelper.TryCreateMatchResult(_completionHelper, itemData.RoslynItem, _filterText,
-                        roslynInitialTriggerKind, roslynFilterReason, _recentItemsManager.GetRecentItemIndex(itemData.RoslynItem), _highlightMatchingPortions, currentIndex,
-                        out var matchResult))
+                    if (threadLocalPatternMatchHelper.Value!.TryCreateMatchResult(itemData.RoslynItem, roslynInitialTriggerKind, roslynFilterReason,
+                        _recentItemsManager.GetRecentItemIndex(itemData.RoslynItem), _highlightMatchingPortions, index, out var matchResult))
                     {
-                        list.Add(matchResult);
+                        lock (_gate)
+                            list.Add(matchResult);
```

Comment:
don't love that inside the parallel we have to lock just to append to a list.  would plinq's Where/Select methods (https://learn.microsoft.com/en-us/dotnet/api/system.linq.parallelenumerable.where?view=net-7.0) method work better?  then you can just enumerate over those results that succeeded.